File: shared variable/c/foo.c
// Compile with `gcc foo.c -Wall -std=gnu99 -lpthread`, or use the makefile
// The executable will be named `foo` if you use the makefile, or `a.out` if you use gcc directly

#include <pthread.h>
#include <stdio.h>

int i = 0;

// Note the return type: void*
void* incrementingThreadFunction(){
    // TODO: increment i 1_000_000 times
    for(int j = 0; j < 1000000; j++) {
        i++;
    }
    return NULL;
}

void* decrementingThreadFunction(){
    // TODO: decrement i 1_000_000 times
    for(int j = 0; j < 1000000; j++) {
        i--;
    }
    return NULL;
}


int main(){
    // TODO: 
    // start the two functions as their own threads using `pthread_create`
    // Hint: search the web! Maybe try "pthread_create example"?
    pthread_t iThread;
    pthread_t dThread;

    if(pthread_create(&iThread, NULL, incrementingThreadFunction, NULL) != 0) {
        perror("pthread_create() error");
    }
    if(pthread_create(&dThread, NULL, decrementingThreadFunction, NULL) != 0) {
        perror("pthread_create() error");
    }
    
    // TODO:
    // wait for the two threads to be done before printing the final result
    // Hint: Use `pthread_join`    

    if (pthread_join(iThread, NULL) != 0) {
        perror("pthread_join() error");
        }

    if (pthread_join(dThread, NULL) != 0) {
        perror("pthread_join() error");
        }


    printf("The magic number is: %d\n", i);
    return 0;
}

File: shared variable/go/foo.go
// Use `go run foo.go` to run your program

package main

import (
	. "fmt"
	"runtime"
)

var i = 0

func incrementing(req chan string, done chan bool) {
	for j := 0; j < 999999; j++ {
		req <- "increment"
	}
	done <- true
}

func decrementing(req chan string, done chan bool) {
	for j := 0; j < 1000000; j++ {
		req <- "decrement"
	}
	done <- true
}

func server(req chan string, resp chan int) {
	for {
		select {
		case value := <-req:
			switch value {
			case "increment":
				i++
			case "decrement":
				i--
			case "get":
				resp <- i
			}
		}
	}
}

func main() {
	// What does GOMAXPROCS do? What happens if you set it to 1?
	runtime.GOMAXPROCS(2)

	req := make(chan string)
	resp := make(chan int)
	done := make(chan bool)

	// TODO: Spawn both functions as goroutines

	go server(req, resp)
	go decrementing(req, done)
	go incrementing(req, done)

	<-done
	<-done

	req <- "get"
	finalValue := <-resp
	Println("The magic number is:", finalValue)
}
